drivers/can/can_sja1000.c

```diff
@@ -0,0 +1,707 @@
+		data->state = CAN_ERROR_WARNING;
+	} else {
+		data->state = CAN_ERROR_PASSIVE;
+	}
+}
+
+void can_sja1000_isr(const struct device *dev)
+{
+	struct can_sja1000_data *data = dev->data;
+	const can_state_change_callback_t cb = data->state_change_cb;
+	void *cb_data = data->state_change_cb_data;
+	enum can_state prev_state = data->state;
+	struct can_bus_err_cnt err_cnt;
+	uint8_t ir;
+
+	ir = can_sja1000_read_reg(dev, CAN_SJA1000_IR);
+
+	if ((ir & CAN_SJA1000_IR_TI) != 0) {
+		can_sja1000_handle_transmit_irq(dev);
+	}
+
+	if ((ir & CAN_SJA1000_IR_RI) != 0) {
+		can_sja1000_handle_receive_irq(dev);
+	}
+
+	if ((ir & CAN_SJA1000_IR_EI) != 0) {
+		can_sja1000_handle_error_warning_irq(dev);
+	}
+
+	if ((ir & CAN_SJA1000_IR_EPI) != 0) {
+		can_sja1000_handle_error_passive_irq(dev);
+	}
+
+	if (prev_state != data->state && cb != NULL) {
+		err_cnt.rx_err_cnt = can_sja1000_read_reg(dev, CAN_SJA1000_RXERR);
+		err_cnt.tx_err_cnt = can_sja1000_read_reg(dev, CAN_SJA1000_TXERR);
+		cb(dev, data->state, err_cnt, cb_data);
+	}
+}
+
+int can_sja1000_init(const struct device *dev)
+{
+	const struct can_sja1000_config *config = dev->config;
+	struct can_sja1000_data *data = dev->data;
+	struct can_timing timing;
+	int err;
+
+	__ASSERT_NO_MSG(config->read_reg != NULL);
+	__ASSERT_NO_MSG(config->write_reg != NULL);
+
+	if (config->phy != NULL) {
+		if (!device_is_ready(config->phy)) {
+			LOG_ERR("CAN transceiver not ready");
+			return -ENODEV;
+		}
+	}
+
+	k_mutex_init(&data->mod_lock);
+	k_sem_init(&data->tx_idle, 1, 1);
+	k_sem_init(&data->tx_done, 0, 1);
+
+	data->state = CAN_ERROR_ACTIVE;
+
+	/* See NXP SJA1000 Application Note AN97076 (figure 12) for initialization sequence */
+
+	/* Enter reset mode */
+	err = can_sja1000_enter_reset_mode(dev);
+	if (err != 0) {
+		return err;
+	}
+
+	/* Set PeliCAN mode */
+	can_sja1000_write_reg(dev, CAN_SJA1000_CDR, config->cdr | CAN_SJA1000_CDR_CAN_MODE);
+
+	/* Set up acceptance code and mask to match any frame (software filtering) */
+	can_sja1000_write_reg(dev, CAN_SJA1000_ACR0, 0x00);
+	can_sja1000_write_reg(dev, CAN_SJA1000_ACR1, 0x00);
+	can_sja1000_write_reg(dev, CAN_SJA1000_ACR2, 0x00);
+	can_sja1000_write_reg(dev, CAN_SJA1000_ACR3, 0x00);
+
+	can_sja1000_write_reg(dev, CAN_SJA1000_AMR0, 0xFF);
+	can_sja1000_write_reg(dev, CAN_SJA1000_AMR1, 0xFF);
+	can_sja1000_write_reg(dev, CAN_SJA1000_AMR2, 0xFF);
+	can_sja1000_write_reg(dev, CAN_SJA1000_AMR3, 0xFF);
+
+	/* Calculate initial timing parameters */
+	data->sjw = config->sjw;
+	timing.sjw = CAN_SJW_NO_CHANGE;
+
+	if (config->sample_point != 0) {
+		err = can_calc_timing(dev, &timing, config->bitrate, config->sample_point);
+		if (err == -EINVAL) {
+			LOG_ERR("bitrate/sample point cannot be met (err %d)", err);
+			return err;
+		}
+
+		LOG_DBG("initial sample point error: %d", err);
+	} else {
+		timing.prop_seg = 0;
+		timing.phase_seg1 = config->phase_seg1;
+		timing.phase_seg2 = config->phase_seg2;
+
+		can_calc_prescaler(dev, &timing, config->bitrate);
```

Comment:
```suggestion
		err = can_calc_prescaler(dev, &timing, config->bitrate);
```

Reply:
Thank you @martinjaeger :-) Added the missing return value check.